### ISSUE TITLE
ActionScheduler post to posts table instead of dedicated tables

### ISSUE THREAD

**Issue by ruggedoutdoor** [Prospress/action-scheduler#812]
Hello,
I did a DB and site restore about a week ago and since then the default table that actions are posting to is the _posts table which is an issue for order increments etc.  Is there a simple way to switch it back to the dedicated _actionscheduler_ tables again?  
WooCommerce Status page is showing this.  
![Screen Shot 2022-04-12 at 7 28 28 AM](https://user-images.githubusercontent.com/11505397/162950416-a93f794f-37c0-4961-a992-1d463fd96c32.png)

**Comment by barryhughes** [comment]
Hi @ruggedoutdoor,

That's unusual.

```php
add_filter( 'action_scheduler_store_class', function () {
    return ActionScheduler_DBStore::class;
}, 1000 );
```

Some code like the above (which you might place in a custom plugin, or mu-plugin) could be used to force use of the custom table store. However, there's not enough information in your report for us to tell if this would be safe. Some things you may wish to check on first:

- The value of the `action_scheduler_migration_status` option.
- If the expected tables (ie, `wp_actionscheduler_actions` etc) actually exist.

**Comment by ruggedoutdoor** [comment]
Hi @barryhughes:
action_scheduler_migration_status is not found in the DB table after the restore I mentioned.  Did a search.  
Also checked for what options exist with %schedule% and attached those.  
The expected actionscheduler tables do exists and used to have data writing to them before things got out of hand and I had to restore the site and DB. Possible that something was not set back after the restore?  It seems like it may be a simple setting to fix in options table.  Not sure of the architecture of this module.  Thank you!
![Screen Shot 2022-04-15 at 3 31 06 PM](https://user-images.githubusercontent.com/11505397/163624661-5d3393f4-8114-4d75-9685-e179be19d8d5.png)

**Comment by ruggedoutdoor** [comment]
Worth noting that I am only using the core adoption of actionscheduler and not the specific plugin.  Never have.

**Comment by barryhughes** [comment]
I think the next thing to try is restoring that option. It looks like you have PhpMyAdmin available, so you could use that to run the following query:

```sql
INSERT INTO gx_options ( 
	option_name, 
	option_value 
)
VALUES (
	"action_scheduler_migration_status",
	"complete"
)
```

Does that help?

**Comment by ruggedoutdoor** [comment]
That seems to have done the trick.  ActionScheduler now knows to use its own tables - I see the actions writing to it and not to posts table anymore.  Is it safe to cleanup the posts table by deleting all records with post_type = "scheduled-action"?

**Comment by peterfabian** [comment]
Hi @ruggedoutdoor , 

It should be fine to clean up the posts with post_type='scheduled-action' (please take a backup first just to be safe).

**Comment by ruggedoutdoor** [comment]
The "﻿action_scheduler_migration_status" option disappeared from gx_options table again.  Do you know why this would then start the writing to the posts table again?

**Comment by barryhughes** [comment]
Anytime a plugin is deactivated, that option is [intentionally deleted](https://github.com/woocommerce/action-scheduler/blob/3.4.0/classes/ActionScheduler_DataController.php#L67-L74). As far as you are aware, could plugin deactivation (could be any plugin) coincide with the removal of the option?

> Do you know why this would then start the writing to the posts table again?

That would seem unexpected. Somewhere near the top of the System Status Report (found in the **WooCommerce ▸ Status** screen) you should see information about the Action Scheduler package:

<img width="1053" alt="ssr-packages-info" src="https://user-images.githubusercontent.com/3594411/165242652-cae8bd13-2a1d-4e03-b72c-66407afd6170.png">

Can you confirm what you see there?

**Comment by ruggedoutdoor** [comment]
Okay what you mention about plugins removal makes sense.  Last night I only updated some plugins and it caused it maybe?
Here is what I show in the WP Env section:

Before option insert again:

After setting the option again it is writing to the actions tables again:

Jim Williams
***@***.***
321-236-3662 PBX
407-920-2751 DIRECT

**Comment by ruggedoutdoor** [comment]
Also I just sent some path info in my last update.  Will you ensure that our conversation is private on Github?
Thank you.

Jim Williams
***@***.***
321-236-3662 PBX
407-920-2751 DIRECT

**Comment by barryhughes** [comment]
> Also I just sent some path info in my last update.  Will you ensure that our conversation is private on Github?

We can edit replies, but the conversation is public (that said, I don't see the path info you are referring to—I'm not sure if you replied by email but perhaps it was in an attachment that did not come through).

What I'm really interested to know is if the Action Scheduler package comes from WooCommerce or not. In the screenshot I shared earlier, you can see this because the path ends in:

`plugins/woocommerce/packages/action-scheduler`

I'm mostly interested to learn if in your case (especially when the problem takes place) it is something else, so I don't really need the *full* path if that's sensitive.

**Comment by ruggedoutdoor** [comment]
Odd the email reply must have left my images out.  Replying through github directly to attach.  
Just copying and pasting that line from the System Status.  
References a library inside of a plugin ip-locator which is odd.  I do not even need that plugin active.  

Action Scheduler package: |   | 3.4.0 /var/www/vitalbody.com/wp-content/plugins/ip-locator/includes/libraries/action-scheduler

![Screen Shot 2022-04-26 at 12 19 56 PM](https://user-images.githubusercontent.com/11505397/165346635-dfd6d198-524d-43e6-9544-e02267d24093.png)

**Comment by barryhughes** [comment]
We took a look at IP Locator and it does have some code that essentially tells Action Scheduler [it cannot migrate to the DB Store](https://plugins.trac.wordpress.org/browser/ip-locator/tags/3.5.0/ip-locator.php#L41). That would seem like a probable cause of the problem (we're less sure of the reason it does this). 

Given you noted not requiring it, could you remove it or at least temporarily deactivate it and see if things consistently work as expected?

**Comment by ruggedoutdoor** [comment]
Thank you for the guidance!

I deactivated that IP Locator plugin and it automatically switched the actionscheduler mode to ActionScheduler_HybridStore. I reapplied the insert statement to set it back to the expected mode and we are looking good again with ActionScheduler_DBStore:
INSERT INTO gx_options (
	option_name,
	option_value
)
VALUES (
	"action_scheduler_migration_status",
	"complete"
)
The action scheduler package path is also reflecting /wp-content/plugins/woocommerce/packages/action-scheduler as it should after deactivating that plugin in question. I will monitor it through the day.

The DB took a bit of a hit but flattened out after about 45 minutes

Jim
